File: backend/app/models/review.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Integer,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel, TimestampMixin

if TYPE_CHECKING:
    from app.models.booking import Booking
    from app.models.master import MasterProfile
    from app.models.user import User


class Review(BaseModel, TimestampMixin):
    __tablename__ = "reviews"

    booking_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("bookings.id"),
        unique=True,
        nullable=False,
    )
    client_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=False,
    )
    master_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("master_profiles.id"),
        nullable=False,
    )
    rating: Mapped[int] = mapped_column(Integer, nullable=False)
    text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    master_reply: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_visible: Mapped[bool] = mapped_column(
        Boolean, server_default="true", nullable=False
    )

    # Relationships
    booking: Mapped["Booking"] = relationship(
        "Booking", back_populates="review"
    )
    client: Mapped["User"] = relationship("User", back_populates="reviews")
    master: Mapped["MasterProfile"] = relationship(
        "MasterProfile", back_populates="reviews"
    )

    __table_args__ = (
        CheckConstraint("rating >= 1 AND rating <= 5", name="ck_reviews_rating_range"),
    )

    def __repr__(self) -> str:
        return f"<Review {self.id} rating={self.rating}>"

File: mobile/lib/shared/models/review.dart
class Review {
  final String id;
  final String? bookingId;
  final String? clientId;
  final String? clientName;
  final String? clientAvatar;
  final String? masterId;
  final int rating;
  final String? text;
  final String? masterReply;
  final DateTime createdAt;

  Review({
    required this.id,
    this.bookingId,
    this.clientId,
    this.clientName,
    this.clientAvatar,
    this.masterId,
    required this.rating,
    this.text,
    this.masterReply,
    required this.createdAt,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['id'] as String,
      bookingId: json['booking_id'] as String?,
      clientId: json['client_id'] as String?,
      clientName: json['client_name'] as String?,
      clientAvatar: json['client_avatar'] as String?,
      masterId: json['master_id'] as String?,
      rating: json['rating'] as int,
      text: json['text'] as String?,
      masterReply: json['master_reply'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
